### PR TITLE
[Enhancement] Delete files with retry in shared data mode

### PR DESCRIPTION
**Pull request by sduzh**
## Why I'm doing:
Deletion fails after qps exceeds object store qps limit.

## What I'm doing:
Add delete retry mechanism.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by sduzh** [comment]
Try to fix https://github.com/StarRocks/starrocks/issues/35567

File: be/src/storage/lake/vacuum.cpp [hunk 1]
```diff
@@ -72,6 +73,33 @@ std::future<Status> completed_future(Status value) {
     return p.get_future();
 }
 
+bool should_retry(const Status& st, int64_t attempted_retries) {
+    if (attempted_retries >= config::lake_vacuum_retry_max_attempts) {
+        return false;
+    }
+    Slice message = st.message();
```
Comment (kevincai) [inline be/src/storage/lake/vacuum.cpp:80]: mapped slow-down request limit to ResourceExhaustedError, may check this type and determine retry or not.

File: be/src/storage/lake/vacuum.cpp [hunk 2]
```diff
@@ -72,6 +73,33 @@ std::future<Status> completed_future(Status value) {
...
+}
+
+int64_t calculate_retry_delay(int64_t attempted_retries) {
+    int64_t min_delay = config::lake_vacuum_retry_min_delay;
+    return min_delay * (1 << attempted_retries);
+}
+
+Status delete_files_with_retry(FileSystem* fs, const std::vector<std::string>& paths) {
+    for (int64_t attempted_retries = 0; /**/; attempted_retries++) {
+        auto st = fs->delete_files(paths);
```
Comment (kevincai) [inline be/src/storage/lake/vacuum.cpp:91]: a little afraid of these batch deletes, some of them are deleted and some of them are not, will this batch retry causes qps amplified for those already deleted files?

Comment (sduzh) [reply be/src/storage/lake/vacuum.cpp:91]: Maybe, but the starlet does not tell which files have been deleted, there is nothing we can do here.

Comment (sduzh) [reply be/src/storage/lake/vacuum.cpp:94]: I'll submit another patch to limit the batch size.

File: be/src/common/config.h [hunk 3]
```diff
@@ -913,6 +913,9 @@ CONF_mInt64(experimental_lake_wait_per_get_ms, "0");
 CONF_mInt64(experimental_lake_wait_per_delete_ms, "0");
 CONF_mInt64(lake_publish_version_slow_log_ms, "1000");
 CONF_mBool(lake_enable_publish_version_trace_log, "false");
+CONF_mString(lake_vacuum_retry_pattern, "*request rate*");
+CONF_mInt64(lake_vacuum_retry_max_attempts, "5");
+CONF_mInt64(lake_vacuum_retry_min_delay, "10");
```
Comment (wyb) [inline be/src/common/config.h:918]:
```suggestion
CONF_mInt64(lake_vacuum_retry_min_delay_ms, "10");
```

File: be/src/storage/lake/vacuum.cpp [hunk 4]
```diff
@@ -72,6 +73,36 @@ std::future<Status> completed_future(Status value) {
...
+        if (!st.ok() && should_retry(st, attempted_retries)) {
+            int64_t delay = calculate_retry_delay(attempted_retries);
+            LOG(WARNING) << "Fail to delete: " << st << " will retry after " << delay << "ms";
+            std::this_thread::sleep_for(std::chrono::milliseconds(delay));
+        } else {
+            LOG_IF(WARNING, !st.ok()) << "Fail to delete: " << st;
```
Comment (wyb) [inline be/src/storage/lake/vacuum.cpp:100]: duplicated log with line 132

File: be/src/common/config.h (hunk 3 above)
Comment (sduzh) [reply be/src/common/config.h]: Updated, PTAL @wyb

**Review by wyb — APPROVED** [review]

**Review by kevincai — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Pull request merged**